### TTC Bus Delays EDA

The 1st step is to download all of the data about bus delays through their API.

We first print all of the files in this package and then download them in this folder.

In [1]:
import requests

# Toronto Open Data is stored in a CKAN instance. It's APIs are documented here:
# https://docs.ckan.org/en/latest/api/

# To hit our API, you'll be making requests to:
base_url = "https://ckan0.cf.opendata.inter.prod-toronto.ca"

# Datasets are called "packages". Each package can contain many "resources"
# To retrieve the metadata for this package and its resources, use the package name in this page's URL:
url = base_url + "/api/3/action/package_show"
params = { "id": "ttc-bus-delay-data"}
package = requests.get(url, params = params).json()

# To get resource data:
for idx, resource in enumerate(package["result"]["resources"]):
    print('{idx}, {name}, {resource}'.format(idx=idx, name=resource['name'], resource=resource))
    # To get metadata for non datastore_active resources:
    if not resource["datastore_active"]:
        url = base_url + "/api/3/action/resource_show?id=" + resource["id"]
        resource_metadata = requests.get(url).json()
        print(resource_metadata)
        # From here, you can use the "url" attribute to download this file

0, ttc-bus-delay-data-readme, {'cache_last_updated': None, 'cache_url': None, 'created': '2019-07-23T18:10:56.253151', 'datastore_active': False, 'format': 'XLSX', 'hash': '', 'id': '71bb1283-7388-4d23-aa4e-0f393a80abce', 'is_datastore_cache_file': False, 'is_preview': 'False', 'last_modified': '2022-04-05T19:21:58', 'metadata_modified': '2022-04-12T18:45:52.481023', 'mimetype': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', 'mimetype_inner': None, 'name': 'ttc-bus-delay-data-readme', 'package_id': 'e271cdae-8788-4980-96ce-6a5c95bc6618', 'position': 0, 'resource_type': None, 'revision_id': 'e90c3e59-7576-4f54-9cb0-6d4375bcc235', 'size': 12839, 'state': 'active', 'url': 'https://ckan0.cf.opendata.inter.prod-toronto.ca/dataset/e271cdae-8788-4980-96ce-6a5c95bc6618/resource/71bb1283-7388-4d23-aa4e-0f393a80abce/download/ttc-bus-delay-data-readme.xlsx', 'url_type': 'upload'}
{'help': 'https://ckan0.cf.opendata.inter.prod-toronto.ca/api/3/action/help_show?name=resource_s

In [2]:
for idx, resource in enumerate(package["result"]["resources"]):
    xlsx_url = resource["url"]
    response = requests.get(xlsx_url)
    
    if response.status_code == 200:
        # Assuming you want to save the XLSX content to a file
        xlsx_filename = f"{resource['name']}.xlsx"
        with open(xlsx_filename, "wb") as file:
            file.write(response.content)
        print(f"XLSX resource downloaded successfully. Check '{xlsx_filename}' file.")
    else:
        print("Failed to download the XLSX resource.")
        break  # Exit the loop if there was an error

XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-readme.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2014.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2015.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2016.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2017.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2018.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2019.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2020.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2021.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2022.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2023.xlsx' file.
XLSX resource downloaded successfully. Check 'ttc-bus-delay-data-2024.xlsx' file.


### Importing packages for EDA

In [3]:
import pandas as pd
import numpy as np

In [4]:
#sometimes it's more convenient to look at DFs with these settings
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

### Readme

In [5]:
df_readme = pd.read_excel('ttc-bus-delay-data-readme.xlsx')
df_readme

Field Name  \
0  Report Date   
1        Route   
2         Time   
3          Day   
4     Location   
5     Incident   
6    Min Delay   
7      Min Gap   
8    Direction   
9      Vehicle   

                                                                                                                                                                                                                              Description  \
0                                                                                                                                                                          The date (YYYY/MM/DD) when the delay-causing incident occurred   
1                                                                                                                                                                                                             The number of the bus route   
2                                                                                                                                                                      The time (hh:mm:ss AM/PM) when the delay-causing incident occurred   
3                                                                                                                                                                                                                     The name of the day   
4                                                                                                                                                                                              The location of the delay-causing incident   
5                                                                                                                                                                                           The description of the delay-causing incident   
6                                                                                                                                                                            The delay, in minutes, to the schedule for the following bus   
7                                                                                                                                                           The total scheduled time, in minutes, from the bus ahead of the following bus   
8  The direction of the bus route where B,b or BW indicates both ways. (On an east west route, it includes both east and west)                                           NB - northbound, SB - southbound, EB - eastbound, WB - westbound   
9                                                                                                                                                                                                                          Vehicle number   

               Example  
0  2017-06-20 00:00:00  
1                   51  
2             00:35:00  
3               Monday  
4   York Mills Station  
5           Mechanical  
6                   10  
7                   20  
8                    N  
9                 1057

In [6]:
df = pd.read_excel('ttc-bus-delay-data-2022.xlsx')
df.head(20)

Date Route   Time       Day                Location  \
0  2022-01-01   320  02:00  Saturday        YONGE AND DUNDAS   
1  2022-01-01   325  02:00  Saturday  OVERLEA AND THORCLIFFE   
2  2022-01-01   320  02:00  Saturday       YONGE AND STEELES   
3  2022-01-01   320  02:07  Saturday       YONGE AND STEELES   
4  2022-01-01   320  02:13  Saturday       YONGE AND STEELES   
5  2022-01-01   363  02:16  Saturday           KING AND SHAW   
6  2022-01-01    96  02:18  Saturday         HUMBERLINE LOOP   
7  2022-01-01   320  02:38  Saturday       STEELES AND YONGE   
8  2022-01-01   320  02:55  Saturday       YONGE AND STEELES   
9  2022-01-01   300  03:18  Saturday         KENNEDY STATION   
10 2022-01-01   300  03:32  Saturday        BLOOR AND INDIAN   
11 2022-01-01    47  03:34  Saturday  LANSDOWNE AND ST CLAIR   
12 2022-01-01    45  03:52  Saturday   DANFORTH AND DANFORTH   
13 2022-01-01    32  04:21  Saturday        EGLINTON STATION   
14 2022-01-01    32  04:39  Saturday       YONGE AND BERWICK   
15 2022-01-01    39  04:42  Saturday     FINCHDENE AND FINCH   
16 2022-01-01    32  04:53  Saturday        RENFORTH STATION   
17 2022-01-01    53  04:58  Saturday     STEELES AND BAYVIEW   
18 2022-01-01    29  05:01  Saturday   DUFFERIN AND LAWRENCE   
19 2022-01-01   334  05:02  Saturday     FINCHDENE AND FINCH   

                 Incident  Min Delay  Min Gap Direction  Vehicle  
0           General Delay          0        0       NaN     8531  
1               Diversion        131      161         W     8658  
2   Operations - Operator         17       20         S        0  
3   Operations - Operator          4       11         S        0  
4   Operations - Operator          4        8         S        0  
5   Operations - Operator         30       60       NaN        0  
6                Security          0        0         N     3536  
7   Operations - Operator          4        8       NaN        0  
8   Operations - Operator          4        8       NaN        0  
9      Emergency Services          0        0         E     8094  
10               Security         17       34         E     8452  
11  Operations - Operator         15       30         S        0  
12  Operations - Operator         30       60         W     1325  
13  Operations - Operator         16       33       NaN     1130  
14     Emergency Services          0        0         S     1267  
15  Operations - Operator         30       60         W        0  
16  Operations - Operator         30        0         E     1073  
17               Security         30       60         E     3299  
18  Operations - Operator         10       20       NaN     9149  
19  Operations - Operator         33        0         W     8744

In [7]:
df.tail(20)

Date Route   Time       Day                Location  \
58687 2022-12-31   121  20:41  Saturday           FRONT AND BAY   
58688 2022-12-31    36  20:53  Saturday        FINCH AND WESTON   
58689 2022-12-31   100  21:23  Saturday     WYNFORD AND DONMILS   
58690 2022-12-31    20  21:30  Saturday         KENNEDY STATION   
58691 2022-12-31    20  21:32  Saturday    DANFORTH AND MIDLAND   
58692 2022-12-31     7  21:40  Saturday     BATHURST AND WILSON   
58693 2022-12-31    36  21:45  Saturday  FINCH AND MARTIN GROVE   
58694 2022-12-31    53  21:45  Saturday           FINCH STATION   
58695 2022-12-31   169  21:48  Saturday  SCARBOROUGH CENTRE STA   
58696 2022-12-31   927  21:50  Saturday  FINCH AND MARTIN GROVE   
58697 2022-12-31     7  22:06  Saturday   BATHURST AND GREENWIN   
58698 2022-12-31    47  22:13  Saturday        YORKDALE STATION   
58699 2022-12-31    63  22:51  Saturday       OSSINGTON STATION   
58700 2022-12-31    72  23:12  Saturday     QUEENS QUAY AND BAY   
58701 2022-12-31    68  23:16  Saturday  ANGUS GLEN COMMUNITY C   
58702 2022-12-31    32  23:21  Saturday   EGLINTON AND EDITH (W   
58703 2022-12-31   113  23:28  Saturday      DANFORTH AND DAWES   
58704 2022-12-31    72  23:56  Saturday  LOWER SHERBOURNE AND L   
58705 2022-12-31    22  00:17  Saturday            BINGHAM LOOP   
58706 2022-12-31     7  00:28  Saturday        BATHURST STATION   

                    Incident  Min Delay  Min Gap Direction  Vehicle  
58687  Operations - Operator         20       40         E     8455  
58688  Operations - Operator         11       22         E     8906  
58689             Mechanical          9       18         S     8820  
58690  Cleaning - Unsanitary         20       40         W     8551  
58691  Cleaning - Unsanitary         20       40         S     8551  
58692  Operations - Operator          9       18         N     8360  
58693              Diversion          0        0         W        0  
58694  Cleaning - Unsanitary         19       39         W     3226  
58695  Cleaning - Unsanitary         30       30         N     3151  
58696              Diversion         64       75         W        0  
58697     Utilized Off Route          9       18         S     8340  
58698  Operations - Operator         20       40         S     8184  
58699          General Delay         10       20         N     8376  
58700              Diversion        292      308         E        0  
58701  Cleaning - Unsanitary         33       66         S     9229  
58702             Mechanical          2       24         W     1085  
58703                 Vision         20       40         W     8589  
58704               Security         18       36         N     8493  
58705  Cleaning - Unsanitary         10       20         S     8582  
58706          General Delay         10       20         N     8354

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58707 entries, 0 to 58706
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       58707 non-null  datetime64[ns]
 1   Route      58323 non-null  object        
 2   Time       58707 non-null  object        
 3   Day        58707 non-null  object        
 4   Location   58707 non-null  object        
 5   Incident   58707 non-null  object        
 6   Min Delay  58707 non-null  int64         
 7   Min Gap    58707 non-null  int64         
 8   Direction  48227 non-null  object        
 9   Vehicle    58707 non-null  int64         
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 4.5+ MB


In [9]:
df.describe()

Date     Min Delay       Min Gap       Vehicle
count                          58707  58707.000000  58707.000000  58707.000000
mean   2022-07-03 11:48:01.066993664     20.115353     32.676154   5467.459298
min              2022-01-01 00:00:00      0.000000      0.000000      0.000000
25%              2022-04-06 00:00:00      9.000000     17.000000   1553.000000
50%              2022-07-13 00:00:00     11.000000     22.000000   7959.000000
75%              2022-09-27 00:00:00     20.000000     36.000000   8546.000000
max              2022-12-31 00:00:00    999.000000    999.000000  93561.000000
std                              NaN     48.945121     50.636856   4356.685772

### Notes
1. Delays are substantial - the median is 11 minutes
2. The mean of the delays is way higher than the median, which indicates a lot of outliers which skew the distribution (for example, you already can see from df.tail(20) 292 minutes delay)
3. Lots of missing values in direction - not a problem for us, it's not an important column, so we can just drop it
4. A bit of missing values in route, we can drop these rows if anything or drop the column too
5. The most important columns IMO are date, time, day, location, incident, and min delay

### Next steps
1. To find out how many delays are under 0 or 5 minutes
2. To print all of the incident values and get additional explanation for them if needed
3. To compare all other years of data (to understand if it's the same format)

In [10]:
df[df['Min Delay'] == 0].info()

<class 'pandas.core.frame.DataFrame'>
Index: 2818 entries, 0 to 58693
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       2818 non-null   datetime64[ns]
 1   Route      2585 non-null   object        
 2   Time       2818 non-null   object        
 3   Day        2818 non-null   object        
 4   Location   2818 non-null   object        
 5   Incident   2818 non-null   object        
 6   Min Delay  2818 non-null   int64         
 7   Min Gap    2818 non-null   int64         
 8   Direction  1604 non-null   object        
 9   Vehicle    2818 non-null   int64         
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 242.2+ KB


In [11]:
df[df['Min Delay'] <= 5].info()

<class 'pandas.core.frame.DataFrame'>
Index: 5602 entries, 0 to 58702
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       5602 non-null   datetime64[ns]
 1   Route      5348 non-null   object        
 2   Time       5602 non-null   object        
 3   Day        5602 non-null   object        
 4   Location   5602 non-null   object        
 5   Incident   5602 non-null   object        
 6   Min Delay  5602 non-null   int64         
 7   Min Gap    5602 non-null   int64         
 8   Direction  3868 non-null   object        
 9   Vehicle    5602 non-null   int64         
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 481.4+ KB


Not so many short delays. Maybe we can just drop the rows below a certain threshold (5 minutes, 10 minutes?). It's basically how we define a significant delay

In [12]:
df['Incident'].value_counts()

Incident
Operations - Operator               19583
Mechanical                          16469
Collision - TTC                      3512
Security                             3374
Utilized Off Route                   3240
General Delay                        3218
Diversion                            2883
Emergency Services                   2421
Cleaning - Unsanitary                1561
Investigation                         906
Vision                                708
Road Blocked - NON-TTC Collision      624
Held By                               162
Late Entering Service                  40
Cleaning - Disinfection                 6
Name: count, dtype: int64

The main reason is an operator incident. If we try to predict this with weather, we can leave only the most sensitive to weather reasons (like Mechanical, Diversion, Collision TTC, Emergency Services, etc.)

In [13]:
df['Min Delay'].value_counts().sort_index()

Min Delay
0      2818
1       121
2        62
3       223
4       603
5      1775
6      1799
7      2589
8      3970
9      4568
10     9453
11     2384
12     2910
13     1682
14     1448
15     2951
16     1262
17     1298
18     1391
19      487
20     3896
21      424
22      786
23      675
24     1017
25      842
26      341
27      214
28      186
29       89
30     3001
31       58
32      115
33      102
34       53
35       81
36       75
37       36
38       54
39       39
40      165
41       22
42       33
43       16
44       37
45       66
46       41
47       24
48       23
49       28
50       74
51       16
52       23
53       22
54       23
55       35
56       16
57       23
58       11
59       16
60      171
61       20
62       24
63       23
64       11
65       21
66       17
67       14
68       23
69       14
70       37
71        8
72       15
73        8
74       16
75       22
76       12
77        5
78       14
79        5
80       27
81        6
82    

### Comparing all of the files

First we will look at .info(), then at .describe()

In [15]:
years = np.arange(2014, 2025, 1)
for year in years:
    df = pd.read_excel('ttc-bus-delay-data-{}.xlsx'.format(year))
    print(df.info())
    print('\n')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9822 entries, 0 to 9821
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Report Date  9822 non-null   datetime64[ns]
 1   Route        9822 non-null   int64         
 2   Time         9822 non-null   object        
 3   Day          9822 non-null   object        
 4   Location     9814 non-null   object        
 5   Incident     9822 non-null   object        
 6   Min Delay    9819 non-null   float64       
 7   Min Gap      9819 non-null   float64       
 8   Direction    9734 non-null   object        
 9   Vehicle      8641 non-null   float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(5)
memory usage: 767.5+ KB
None


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6665 entries, 0 to 6664
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Repor

In [17]:
for year in years:
    df = pd.read_excel('ttc-bus-delay-data-{}.xlsx'.format(year))
    print(year, len(df)) #number of delays per year

2014 9822
2015 6665
2016 5326
2017 5300
2018 6969
2019 6743
2020 4282
2021 2832
2022 58707
2023 56207
2024 4764


In [18]:
for year in years:
    df = pd.read_excel('ttc-bus-delay-data-{}.xlsx'.format(year))
    print(df.describe())
    print('\n')

                         Report Date        Route    Min Delay      Min Gap  \
count                           9822  9822.000000  9819.000000  9819.000000   
mean   2014-01-16 01:20:29.321930240    86.089188    12.651288    21.837560   
min              2014-01-01 00:00:00     1.000000    -1.000000     0.000000   
25%              2014-01-07 00:00:00    36.000000     6.000000    10.000000   
50%              2014-01-16 00:00:00    60.000000     9.000000    17.000000   
75%              2014-01-24 00:00:00   105.000000    14.000000    26.000000   
max              2014-01-31 00:00:00   909.000000  1160.000000   730.000000   
std                              NaN    89.551231    27.901961    29.568837   

           Vehicle  
count  8641.000000  
mean   5402.250434  
min       0.000000  
25%    1554.000000  
50%    7401.000000  
75%    7932.000000  
max    9965.000000  
std    3184.568138  


                         Report Date        Route    Min Delay      Min Gap  \
count             

It's weird that the size of dfs range from 3000 to 50000. One of the explanations - covid. Duration of delays increased over the years, but not extremely

In [19]:
for year in years:
    df = pd.read_excel('ttc-bus-delay-data-{}.xlsx'.format(year))
    print(year, df['Incident'].value_counts())
    print('\n')

2014 Incident
Mechanical             5401
Utilized Off Route     1356
Late Leaving Garage    1148
General Delay          1010
Investigation           536
Diversion               267
Emergency Services      104
Name: count, dtype: int64


2015 Incident
Mechanical             2961
Late Leaving Garage    1325
Utilized Off Route     1074
Investigation           480
General Delay           475
Diversion               247
Emergency Services      103
Name: count, dtype: int64


2016 Incident
Mechanical             2342
Late Leaving Garage     864
Utilized Off Route      845
Investigation           486
General Delay           445
Diversion               257
Emergency Services       87
Name: count, dtype: int64


2017 Incident
Mechanical             2164
Late Leaving Garage    1007
Utilized Off Route      975
Investigation           499
General Delay           287
Diversion               262
Emergency Services      106
Name: count, dtype: int64


2018 Incident
Mechanical             2246
Late L

Operations - Operator category appeared only in 2021. Still the difference in observations per year is just too big. Good for discussion - why the numbers of observations differ so much from year to year?